## Откройте и изучите файл

In [1]:
# импорт бибилотеки pandas
import pandas as pd
# импорт метрики качества из библиотеки sklearn
from sklearn.metrics import accuracy_score 
# импорт функции деления выборки из библиотеки sklearn
from sklearn.model_selection import train_test_split 
# импорт дерева решений из библиотеки sklearn
from sklearn.tree import DecisionTreeClassifier
# импорт случайного леса из библиотеки sklearn
from sklearn.ensemble import RandomForestClassifier
# импорт логистической регрессии из библиотеки sklearn
from sklearn.linear_model import LogisticRegression 

In [2]:
# чтение файла с данными и сохранение в df
df = pd.read_csv('users_behavior.csv') 
# получение информации о таблице
display(df.info())
# выведем на экран таблицу
df.head(3)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


None

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0

#### Вывод

Данные получены в готовом виде, предобработка не требуется.

## Разбейте данные на выборки

In [3]:
# выделим из таблицы признаки
features = df.drop(['is_ultra'], axis=1)
# выделим из таблицы целевой признак
target = df['is_ultra']
# деление на обучающую и валидационную выборки
features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=0.4, random_state=57117)
# выведем на экран размер признаков для обучения
display(f'{features_train.shape} features_train')
# выведем на экран размер целей для обучения
display(f'{target_train.shape} target_train')
# деление на обучающую и валидационную выборки
features_test, features_valid, target_test, target_valid = train_test_split(features_test, target_test, test_size=0.5, random_state=57117)
# выведем на экран размер признаков для теста
display(f'{features_test.shape} features_test')
# выведем на экран размер целей для теста
display(f'{target_test.shape} target_test')
# выведем на экран размер признаков для валидации
display(f'{features_valid.shape} features_valid')
# выведем на экран размер целей для валидации
display(f'{target_valid.shape} target_valid')

'(1928, 4) features_train'

'(1928,) target_train'

'(643, 4) features_test'

'(643,) target_test'

'(643, 4) features_valid'

'(643,) target_valid'

#### Вывод

Итогом разбиения исходной таблицы стали следующие выборки:
- Обучающая (60% исходной таблицы df) в количестве 1928 строк
- Тестовая (20% исходной таблицы df) в количестве 643 строки
- Валидационная (20% исходной таблицы df) в количестве 643 строки

## Исследуйте модели

### Решающее дерево

In [4]:
# создадим переменную для записи наилучшей глубины
depth_max = 0
# создадим переменную для запсии лучшего результата
top_ac_score = 0
# создадим цикл для получения нужной глубины
for depth in range(1, 11):
# создадим модель
    model_one = DecisionTreeClassifier(random_state=57117, max_depth=depth)
# обучим модель
    model_one.fit(features_train, target_train)
# найдем предсказания на валидационной выборке
    predictions_valid = model_one.predict(features_valid)
# посчитаем результат на валидационной выборке    
    ac_score = accuracy_score(predictions_valid, target_valid)
# создадим цикл для записи наилучшего результата
    if ac_score > top_ac_score:
# сохраним наилучшее значение глубины
        depth_max = depth
# сохраним наилучшее значение результата
        top_ac_score = ac_score
# выведем на экран все полученные варианты
display(f'Наилучшее значение Accuracy: {top_ac_score}, при глубине модели: {depth_max}') 

'Наилучшее значение Accuracy: 0.7947122861586314, при глубине модели: 7'

#### Вывод

По результатам расчетов модели решающего дерева на валидационной выборке добились показателя Accuracy = 0.794, при глубине модели = 7

### Случайный лес

In [5]:
# создадим пустую переменную для записи модели
best_model_two = None
# создадим пустую переменную для записи результатов 
best_result_two = 0
# создадим пустую переменную для записи количества оценщиков
best_est = 0
# создадим цикл для получения количества "оценщиков" модели
for est in range(1, 11):
# обучим модель с заданным количеством деревьев    
    model_two = RandomForestClassifier(random_state=57117, n_estimators=est) 
# обучим модель на обучающей выборке    
    model_two.fit(features_train, target_train)
# посчитаем качество модели на валидационной выборке    
    predictions_valid_two = model_two.predict(features_valid)
# получим значение результата    
    result_two = accuracy_score(predictions_valid_two, target_valid) 
# зададим условие записи результатов    
    if result_two > best_result_two:
# сохраним наилучшую модель        
        best_model_two = model_two
# сохраним наилучшее значение метрики качества на валидационных данных    
        best_result_two = result_two 
# сохраним лучшее количество оценщиков    
        best_est = est
# выведем результат на экран
display(f'Accuracy наилучшей модели: {best_result_two}, при количестве "деревьев": {best_est}')

'Accuracy наилучшей модели: 0.7931570762052877, при количестве "деревьев": 4'

#### Вывод

По результатам расчетов модели случайного леса на валидационной выборке добились показателя Accuracy = 0.793, при количестве деревьев = 4

### Логистическая регрессия

In [6]:
# создадим модель логистической регресии
model_three = LogisticRegression(random_state=57117, solver='lbfgs', max_iter=1000) 
# обучим модель логистической регресии
model_three.fit(features_train, target_train)
# вычислим предсказания
predictions_valid_three = model_three.predict(features_valid) 
# рассчитаем результат
ac_score_three = accuracy_score(predictions_valid_three, target_valid)
# выведем значение на экран
display(f'Accuracy для логистической регресии: {ac_score_three}')

'Accuracy для логистической регресии: 0.7371695178849145'

#### Вывод

По результатам расчетов модели логистической регрессии на валидационной выборке добились показателя Accuracy = 0.737

## Проверьте модель на тестовой выборке

In [7]:
# рассчитаем предсказания для первой модели
predictions_test_one = model_one.predict(features_test)
# выведем на экран проверку модели на тестовой выборке
display(f'Accuracy модели решающего дерева на тестовой выборке: {accuracy_score(target_test, predictions_test_one)}')
# рассчитаем предсказания для второй модели
predictions_test_two = model_two.predict(features_test)
# выведем на экран проверку модели на тестовой выборке
display(f'Accuracy модели случайного леса на тестовой выборке: {accuracy_score(target_test, predictions_test_two)}')
# рассчитаем предсказания для третьей модели
predictions_test_three = model_three.predict(features_test)
# выведем на экран проверку модели на тестовой выборке
display(f'Accuracy модели логистической регрессии на тестовой выборке: {accuracy_score(target_test, predictions_test_three)}')

'Accuracy модели решающего дерева на тестовой выборке: 0.7884914463452566'

'Accuracy модели случайного леса на тестовой выборке: 0.7978227060653188'

'Accuracy модели логистической регрессии на тестовой выборке: 0.7542768273716952'

#### Вывод

Все три рассмотренные модели показали значение Accuracy на тестовой выборке не ниже 0.75, при том что модель решающего дерева и модель случайного леса также показали метрику качества не ниже 0.75 на валидационной выборке, а модель логистичесской регрессии показала всего лишь 0.737.

## (бонус) Проверьте модели на адекватность

In [8]:
# посчитаем количество единиц в целевом столбце
target_count_one = target_test.count() - target_test.sum()
# посчитаем accuracy модели которая будет прогнозировать только нули
accuracy_bonus = target_count_one / target_test.count()
# выведем значение на экран
display(f'Accuracy модели с доминирующем классом = {accuracy_bonus}')

'Accuracy модели с доминирующем классом = 0.6920684292379471'

#### Вывод

Модель случайного леса, обученная нормально имеет показатель accuracy равный 0,793. Модель предсказывающая только доминирующий класс имеет показатель 0,692. Можем сделать вывод, что наша модель работает нормально.